pkg/cmd/pulumi/stack_migrate.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+func migrateStack(
+	ctx context.Context, opts display.Options,
+	getBackend func(context.Context, display.Options, string) (backend.Backend, error),
+	sourceStack string, version string, targetStack string,
+	targetURL string, secretsProvider string) error {
+
+	// Fetch the current stack and export its deployment
+	s, err := requireStack(ctx, sourceStack, false, opts, false /*setCurrent*/)
+	if err != nil {
+		return err
+	}
+	currentProjectStack, err := loadProjectStack(s)
+	if err != nil {
+		return err
+	}
+
+	var deployment *apitype.UntypedDeployment
+	// Export the latest version of the checkpoint by default. Otherwise, we require that
+	// the backend/stack implements the ability the export previous checkpoints.
+	if version == "" {
+		deployment, err = s.ExportDeployment(ctx)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Check that the stack and its backend supports the ability to do this.
+		be := s.Backend()
+		specificExpBE, ok := be.(backend.SpecificDeploymentExporter)
+		if !ok {
+			return fmt.Errorf("the current backend (%s) does not provide the ability to export previous deployments",
+				be.Name())
+		}
+
+		deployment, err = specificExpBE.ExportDeploymentForVersion(ctx, s, version)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Build decrypter based on the existing secrets provider
+	var decrypter config.Decrypter
+	currentConfig := currentProjectStack.Config
+
+	if currentConfig.HasSecureValue() {
+		dec, decerr := getStackDecrypter(s)
+		if decerr != nil {
+			return decerr
+		}
+		decrypter = dec
+	} else {
+		decrypter = config.NewPanicCrypter()
+	}
```

Comment:
nit because I always like the challenge of eliminating `else`
```suggestion
	getDeployment := s.ExportDeployment
	if version != "" {
		// Check that the stack and its backend supports the ability to export previous deployments
		be := s.Backend()
		specificExpBE, ok := be.(backend.SpecificDeploymentExporter)
		if !ok {
			return fmt.Errorf("the current backend (%s) does not provide the ability to export previous deployments",
				be.Name())
		}
		getDeployment = func(ctx context.Context) (*apitype.UntypedDeployment, error) {
			return specificExpBE.ExportDeploymentForVersion(ctx, s, version)
		}
	}

	deployment, err := getDeployment(ctx)
	if err != nil {
		return err
	}

	// Build decrypter based on the existing secrets provider
	decrypter := config.NewPanicCrypter().(config.Decrypter)
	currentConfig := currentProjectStack.Config

	if currentConfig.HasSecureValue() {
		dec, decerr := getStackDecrypter(s)
		if decerr != nil {
			return decerr
		}
		decrypter = dec
	}
```

Reply:
the current is inline with what these code blocks look like elsewhere so I'm inclined to match the current style. 
In a perfect world go would be a good language with expression based "if" and else blocks would make perfect sense, but alas.